### ISSUE TITLE
Breaks formatting and syntax highlighting on embedded code blocks

### ISSUE THREAD

**Issue by Sokushinbutsu** [mozilla/readability#532]
When using readability on [a site that has embedded code](https://www.taniarascia.com/adding-syntax-highlighting-to-code-snippets/) from gist/codepen or inline code samples surrounded with `<code>` and `<pre>` tags this extension removes syntax highlighting and formatting.

I think that website css should be preserved with code samples and embedding should be allowed from gist, codepen, repl etc.

Embedding docs:
https://help.github.com/en/articles/about-gists#embedding-gists
https://repl.it/site/docs/repls/embed
https://blog.codepen.io/documentation/features/embedded-pens/

**Firefox 65.0.1 on Ubuntu 18.04**

**Comment by gijsk** [comment]
> I think that website css should be preserved

We can't really do this, it would interfere with what reader mode is supposed to do, which is give the user control over how the page displays.

> embedding should be allowed from gist, codepen, repl etc.

It's not about it not being allowed, it's that we strip as much as possible to leave the barebones HTML markup. This still works correctly, but in this case it means syntax highlighting is lost. It's not 100% clear what would be the best way to do this, because people have reinvented the "syntax highlighting on a webpage" wheel approximately 42 bazillion times, slightly differently every time.

Also relevant: https://bugzilla.mozilla.org/show_bug.cgi?id=1140967 .

**Comment by chan-dev** [comment]
Is there any option on what html tags list for this package to allow?

Something like this:
`var article = new Readability(document, {
    allowedList: ['code', 'mark']
}).parse();`

**Comment by gijsk** [comment]
> Is there any option on what html tags list for this package to allow?

No. I think it'd be worth updating the logic in readability to not touch descendants of `code` elements at all. That seems like it'd solve the issue of comments or other items going missing?

As for `mark` tags, I would not expect readability to remove them unless they had classes or identifiers on them that matched some of the regular expressions used to remove extraneous content. Do you have an example of where this happens?

**Comment by chan-dev** [comment]
I just set the `mark` tag as an example here. I believe I haven't tested it yet with a website with a `mark` tag. 

With the `code` element, that's one approach that I feel like an easy fix for code blocks.

This package seems to remove **form elements* but leave the **<label>** tags?
Try to test it with any stackoverflow page and you'll see the form labels are left untouched but form inputs and the like are removed.

**Comment by gijsk** [comment]
> form inputs and the like are removed.

That seems unrelated to this issue. I also can't reproduce form labels being retained (tested on [this page](https://stackoverflow.com/questions/3749231/download-file-using-javascript-jquery)). Please file a separate issue.

**Comment by chan-dev** [comment]
I've created one. here it is  #613

**Comment by twome** [comment]
This problem makes Reader View basically useless for any article featuring code, which is heartbreaking because I use it all the time. Anything in a `<pre>` or `<code>` block _needs_ to be monospaced at the absolute bare minimum. It's even part of the semantic meaning of `<pre>` to leave the formatting exactly as written; changing originally monospaced text to a proportional font goes directly against standards.

Syntax highlighting is of course a can of worms but monospacing is a one-line fix that would have enormous impact.

**Comment by ShahinSorkh** [comment]
I love "Dumb palette-swapping to a Reader View palette" idea.

The problem is, with reader mode as is now, reading articles with code blocks is almost impossible.

**Comment by ShahinSorkh** [comment]
As the bare minimum, you can draw a border around the code blocks and set its font to a monospace font. So that it is distinguishable from the other parts of the content:

(This is a part of mozilla/readability README.md)
![image](https://github.com/mozilla/readability/assets/13497697/032f8518-d685-4966-8168-cac10888b362)

**Comment by twome** [comment]
Can't believe I forgot to mention drawing a border! Yeah, that (or a different background colour that creates a similarly unambiguous boundary) is absolutely essential

**Comment by kravlad** [comment]
Hi guys,

could you share your code with drawing a border and setting the font to a monospace one. I'm not too cool in html and css and it removes language name from code block no matter I try.

Thanks in advance!

**Comment by gijsk** [comment]
The CSS formatting used for `<pre>` or `<code>` in Firefox is not in this repository, so tickets/problems relating strictly to that belong on https://bugzilla.mozilla.org/ . I note that `<code>` blocks already use both fixed width fonts and a border/outline. I also noticed that we regressed `<pre>` tags as a result of some changes to deal with `text/plain` responses. I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1958780 for this and would be happy to mentor someone to create a patch to fix this (avoiding using `<pre>` for `text/plain` (some JS but not hard) and improving the styling of `<pre>` so it goes back to being fixed width fonts - that part is just removing a single CSS rule!) - but it wouldn't be in this repository or github. If you're interested, please comment on the bug (you can sign in with a github account).

This ticket was filed about maintaining syntax highlighting, and any changes to `readability` that would support doing that. At this point, readability already supports a "classes to preserve" list option, and so one potential way to address this issue as filed may be to figure out what common code styling frameworks use to style code, add those classes to that list, and then add CSS in Firefox that styles those classes.

**Comment by kravlad** [comment]
> The CSS formatting used for `\<pre\>` or `<code>` in Firefox is not in this repository, so tickets/problems relating strictly to that belong on https://bugzilla.mozilla.org/ . I note that `<code>` blocks already use both fixed width fonts and a border/outline. I also noticed that we regressed `\<pre\>` tags as a result of some changes to deal with `text/plain` responses. I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1958780 for this and would be happy to mentor someone to create a patch to fix this (avoiding using `\<pre\>` for `text/plain` (some JS but not hard) and improving the styling of `\<pre\>` so it goes back to being fixed width fonts - that part is just removing a single CSS rule!) - but it wouldn't be in this repository or github. If you're interested, please comment on the bug (you can sign in with a github account).
> 
> This ticket was filed about maintaining syntax highlighting, and any changes to `readability` that would support doing that. At this point, readability already supports a "classes to preserve" list option, and so one potential way to address this issue as filed may be to figure out what common code styling frameworks use to style code, add those classes to that list, and then add CSS in Firefox that styles those classes.

Hi Gijs!

"classes to preserve" is exactly what I need.

Thank you for your response!

**Comment by inchei** [comment]
As a workaround, I've tweaked my userContent.css, using [FontWithASyntaxHighlighter](https://github.com/hlotvonen/FontWithASyntaxHighlighter)
```css
.moz-reader-content {
    code p {
        display: contents;
    }
    code:has(span) {
        display: block;
        white-space: pre-wrap;
    }
    code, pre {
        font-family: "FontWithASyntaxHighlighterLightOwl Light", monospace;
    }
}
```
